content/blog/2024/2024-04-28-spring-cloud-sns-sqs-pubsub.md

```diff
@@ -0,0 +1,589 @@
+
+## Subscribing SQS Queue to SNS Topic
+
+Now that we have both of our microservices set up, there's one final piece of the puzzle to connect: **subscribing our SQS queue to our SNS topic**. This will allow the messages published to the SNS topic `user-account-created` to automatically be forwarded to the SQS queue `dispatch-email-notification` for consumption by our subsriber microservice.
+
+To create a subscription between the services, the <a href="https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-configure-subscribe-queue-sns-topic.html" target="_blank">official documentation guide</a> can be referenced.
+
+### Resource Based Policy
+
+Once our subscription has been created, we need to grant our SNS topic, the permission to send messages to our SQS queue. This permission needs to be added to our queue's resource-based policy (Access policy).
+
+You might wonder why this is necessary when we have already granted required IAM permissions to our microservices. The answer lies in the way AWS services communicate with each other. IAM permissions control what actions an IAM user can perform on an AWS resource, **while resource-based policies determine what actions another AWS service can perform on it**. Resource-based policies are attached to an AWS resource (SQS in this context).
+
+In our case, we need to create a resource-based policy on the SQS queue to allow the SNS topic to send messages to it. **Without this policy, even though our microservices have the necessary IAM permissions, the SNS topic will not be able to forward messages to the SQS queue**.
+
+Here is what our SQS resource policy should look like:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "Service": "sns.amazonaws.com"
+      },
+      "Action": "sqs:SendMessage",
+      "Resource": "sqs-queue-arn",
+      "Condition": {
+        "ArnEquals": {
+          "aws:SourceArn": "sns-topic-arn"
+        }
+      }
+    }
+  ]
+}
+```
+
+In this policy, we are granting the SNS service `(sns.amazonaws.com)` permission to perform the `sqs:SendMessage` action on our SQS queue. We specify the ARNs of our SQS queue and SNS topic in the `Resource` and `Condition` field respectively to ensure that only messages from our specific topic are allowed.
+
+Once this resource-based policy is attached to our SQS queue, the SNS topic will be able to forward messages to it, finally completing the setup of our publisher-subscriber architecture.
+
+## Encryption at Rest Using KMS
+
+When dealing with sensitive data, it is recommended to ensure that the data is encrypted not only in transit but also at rest. Encryption at rest, not only enhances the security of our architecture but also makes our life easier when going through HIPAA and PCI-DSS audits.
```

Comment:
Maybe you can explain in more detail why this is data "at rest". SNS and SQS are about sending messages, which is data in transit.

Reply:
Yes, that would be helpful. I've added in the explanation of when the data would be at rest in our solution.